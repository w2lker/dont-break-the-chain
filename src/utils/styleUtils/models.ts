export type cssAttribute = 'unset' | 'null' | number | string | undefined | null;

type TArrayWithLength<TItem, TLength extends number> = [TItem, ...TItem[]] & { length: TLength };
export type BoundAttributeArray = TArrayWithLength<string, 1| 2 | 3 | 4>;
