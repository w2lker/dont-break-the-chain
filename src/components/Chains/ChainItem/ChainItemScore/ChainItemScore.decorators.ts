import withStyles from '@material-ui/core/styles/withStyles';

import ChainItemScore from './ChainItemScore';
import chainItemScoreStyles from './ChainItemScore.styles';

// @ts-ignore
const styledComponent = withStyles(chainItemScoreStyles)(ChainItemScore);

export default styledComponent;
