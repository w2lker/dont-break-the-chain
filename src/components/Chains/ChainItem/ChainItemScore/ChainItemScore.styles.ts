import { styleUtils } from '../../../../utils/styleUtils';

const chainItemScoreStyles = {
  root: {
    ...styleUtils.widthMinMax(70),
    ...styleUtils.setFont(14, '#C2CCD1', 500, '0.1px'),
    textAlign: 'center',
    lineHeight: 24,
  },
};

export default chainItemScoreStyles;