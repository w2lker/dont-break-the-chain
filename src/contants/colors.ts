import colorBlue from '@material-ui/core/colors/blue';
import colorBrown from '@material-ui/core/colors/brown';
import colorCyan from '@material-ui/core/colors/cyan';
import colorDeepPurple from '@material-ui/core/colors/deepPurple';
import colorGreen from '@material-ui/core/colors/green';
import colorIndigo from '@material-ui/core/colors/indigo';
import colorOrange from '@material-ui/core/colors/orange';
import colorPink from '@material-ui/core/colors/pink';
import colorPurple from '@material-ui/core/colors/purple';
import colorTeal from '@material-ui/core/colors/teal';

export type colorsKeys = 'pink' | 'purple' | 'deepPurple' | 'indigo' | 'blue' | 'cyan' | 'teal' | 'green' | 'orange' | 'brown';

export const colors: {[color: string]: any}  = {
  blue: colorBlue,
  brown: colorBrown,
  pink: colorPink,
  cyan: colorCyan,
  deepPurple: colorDeepPurple,
  indigo: colorIndigo,
  green: colorGreen,
  orange: colorOrange,
  purple: colorPurple,
  teal: colorTeal,
};
