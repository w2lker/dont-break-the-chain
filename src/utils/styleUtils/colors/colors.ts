import {bonWarning} from '../helpers/warning/warning';

export interface IRgbColor {
  red: number;
  green: number;
  blue: number;
}

const black: IRgbColor = {
  red: 0,
  green: 0,
  blue: 0,
};

function regularHexToRgb(hex: string): IRgbColor {
  const [red, green, blue] = [0, 2, 4]
    .map(offset => hex.substring(offset, offset + 2))
    .map(hexCh => parseInt(hexCh, 16));
  return { red, green, blue };
}

export function hexToRgbMapped(hex?: string): IRgbColor {
  const warnType = 'HEX to RGB converter';
  if (!hex) {
    return black;
  }
  if (hex.length === 1) {
    return regularHexToRgb(hex.repeat(6));
  }

  if (hex.length === 2) {
    return regularHexToRgb(hex.repeat(3));
  }

  if (hex.length === 3) {
    const [r, g, b] = hex.split('') as string[];
    return regularHexToRgb(`${r}${r}${g}${g}${b}${b}`);
  }

  if (hex.length === 4 || hex.length === 5) {
    const hex3 = hex.slice(0, 3);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex3} value`, warnType);
    return hexToRgbMapped(hex3);
  }

  if (hex.length > 6) {
    const hex6 = hex.slice(0, 6);
    // tslint:disable-next-line:no-console
    bonWarning(`can't verify converting from HEX to RGB — received HEX value with ${hex.length} digits. It is parsed as ${hex6} value`, warnType);
    return regularHexToRgb(hex6);
  }

  return regularHexToRgb(hex);
}
