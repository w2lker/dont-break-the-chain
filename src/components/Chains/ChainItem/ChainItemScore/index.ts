import ChainItemScore from './ChainItemScore';
export default ChainItemScore;
