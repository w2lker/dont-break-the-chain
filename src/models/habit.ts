import {List} from 'immutable';
import {colorsKeys} from "../contants/colors";

export const habitStatus = {
  incomplete: 0,
  complete: 1,
  paused: 2,
  start: 3,
  end: 4,
  incompleteToday: 5
};

export interface IHabitDate {
  date?: any;
  status?: string;
  score?: number;
}

export interface IHabit {
  id?: number;
  name?: string;
  color?: colorsKeys;
  icon?: string;
  currentChain?: number;
  longestChain?: number;
  dates?: List<IHabitDate>;
}