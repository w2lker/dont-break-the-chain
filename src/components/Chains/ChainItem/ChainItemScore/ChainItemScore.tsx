import classNames from 'classnames';
import React from 'react';

import { colors, colorsKeys } from '../../../../contants/colors';
import { habitStatus, habitStatusesType } from '../../../../models/habit';

export interface IChainItemScoreProps {
  status: habitStatusesType;
  color: colorsKeys;
  value: number;
}

const ChainItemScore: React.FC<IChainItemScoreProps> = (props) => {
  const { status, value } = props;

  const scoreClass = classNames({
    'score-wrapper': true,
    hidden: (status === habitStatus.incomplete) || (status === habitStatus.incompleteToday) || (!value),
    bright: value > 4,
  });

  const colorBase = colors[props.color || 'blue'];
  const color = colorBase[value > 4 ? 500 : 400];

  const outputValue = value ? `+${value}` : null;

  return (
    <span
      className={scoreClass}
      style={{ color }}
    >
      {outputValue}
    </span>
  );
};

export default ChainItemScore;
