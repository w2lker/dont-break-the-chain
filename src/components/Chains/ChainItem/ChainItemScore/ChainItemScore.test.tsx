import { mount, shallow } from 'enzyme';
import React from 'react';

import ChainItemScore, { IChainItemScoreProps } from './ChainItemScore';
import ChainItemScoreDecorated from './ChainItemScore.decorators';
import chainItemScoreStyles from './ChainItemScore.styles';

import { colors } from '../../../../contants/colors';
import { habitStatus } from '../../../../models/habit';

const classes = {
  root: 'chain-item-score-root',
};

describe('Chain Item Score', () => {
  const emptyProps: IChainItemScoreProps = {
    classes,
      // @ts-ignore
    status: null,
      // @ts-ignore
    color: null,
      // @ts-ignore
    value: null,
  };

  const sampleProps: IChainItemScoreProps = {
    classes,
      // @ts-ignore
    status: habitStatus.complete,
    // @ts-ignore
    color: 'green',
    value: 3,
  };

  it('matches snapshot', () => {
    const component = mount(<ChainItemScore {...sampleProps} />);
    expect(component.debug()).toMatchSnapshot();
  });

  describe('Empty props rendering', () => {
    const component = shallow(<ChainItemScore {...emptyProps} />);
    it('assigns hidden classname', () => {
      expect(component.find('span').hasClass(classes.root)).toBe(true);
      expect(component.find('span').hasClass('hidden')).toBe(true);
      expect(component.find('span').hasClass('bright')).toBe(false);
    });

    it('assigns default color', () => {
      const emptyStateStyles = {
        color: colors.blue[400],
      };
      const spanStyles = component.find(`.${classes.root}`).props().style;
      expect(spanStyles).toEqual(emptyStateStyles);
    });

    it('hides score value output', () => {
      expect(component.find(`.${classes.root}`).text()).toBe('');
    });
  });

  describe('Sample props rendering', () => {
    const component = shallow(<ChainItemScore {...sampleProps}/>);
    it("doesn't assign hidden class", () => {
      expect(component.find('span').hasClass(classes.root)).toBe(true);
      expect(component.find('span').hasClass('hidden')).toBe(false);
    });
    it('outputs score value', () => {
      expect(component.find(`.${classes.root}`).text()).toBe(`+${sampleProps.value}`);
    });
    it('assigns correct color', () => {
      const sampleStyle = {
        color: colors[sampleProps.color][400],
      };
      expect(component.find(`.${classes.root}`).props().style).toEqual(sampleStyle);
    });
  });

  describe('Rendering with high score value', () => {
    const newProps = {
      ...sampleProps,
      value: 15,
    };
    const component = shallow(<ChainItemScore {...newProps}/>);
    it('assigns accent class', () => {
      expect(component.find('span').hasClass(classes.root)).toBe(true);
      expect(component.find('span').hasClass('hidden')).toBe(false);
      expect(component.find('span').hasClass('bright')).toBe(true);
    });

    it('assigns correct color', () => {
      const sampleStyle = {
        color: colors[sampleProps.color][500],
      };
      expect(component.find(`.${classes.root}`).props().style).toEqual(sampleStyle);
    });
  });
});

describe('Chain Item Score Styles', () => {
  it('should contain principle fields', () => {
    expect(chainItemScoreStyles).toBeDefined();
    expect(chainItemScoreStyles.root).toBeDefined();
  });
});

describe('Chain Item Score Decorated', () => {
  const props: IChainItemScoreProps = {
    // @ts-ignore
    status: habitStatus.complete,
    // @ts-ignore
    color: 'green',
    value: 3,
  };
  it('provides styled classes from decorators', () => {
    const component = mount(<ChainItemScoreDecorated {...props} />);
    // @ts-ignore
    const assignedClasses = component.find('ChainItemScore').props().classes;
    expect(assignedClasses).toBeDefined();
    expect(assignedClasses.root).toBeDefined();
  });
});
