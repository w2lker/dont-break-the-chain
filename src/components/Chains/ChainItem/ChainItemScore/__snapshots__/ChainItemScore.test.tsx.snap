// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chain Item Score matches snapshot 1`] = `
"<ChainItemScore status=\\"habitStatus/completed\\" color=\\"green\\" value={3}>
  <span className=\\"score-wrapper\\" style={{...}}>
    +3
  </span>
</ChainItemScore>"
`;
